drivers/sensor/lsm6dso/lsm6dso.c

```diff
@@ -330,6 +355,46 @@ static int lsm6dso_sample_fetch_gyro(const struct device *dev)
 	return 0;
 }
 
+#if defined(CONFIG_LSM6DSO_ENABLE_FIFO)
+static int lsm6dso_sample_fetch_fifo(const struct device *dev)
+{
+	const struct lsm6dso_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm6dso_data *data = dev->data;
+	uint8_t buff[6];
+	uint16_t num_data_fifo;
+	lsm6dso_fifo_tag_t data_tag;
+
+	lsm6dso_fifo_data_level_get(ctx, &num_data_fifo);
+	for (int i = 0; i < num_data_fifo; i++) {
+		lsm6dso_fifo_sensor_tag_get(ctx, &data_tag);
+		if (lsm6dso_fifo_out_raw_get(ctx, buff) < 0) {
+			LOG_DBG("Failed to read sample");
+			return -EIO;
+		}
+		if (data_tag == LSM6DSO_GYRO_NC_TAG) {
+			data->gyro[0] = (int16_t)buff[1];
+			data->gyro[0] = (data->gyro[0] * 256) + (int16_t)buff[0];
+			data->gyro[1] = (int16_t)buff[3];
+			data->gyro[1] = (data->gyro[1] * 256) + (int16_t)buff[2];
+			data->gyro[2] = (int16_t)buff[5];
+			data->gyro[2] = (data->gyro[2] * 256) + (int16_t)buff[4];
+		} else if (data_tag == LSM6DSO_XL_NC_TAG) {
+			data->acc[0] = (int16_t)buff[1];
+			data->acc[0] = (data->acc[0] * 256) + (int16_t)buff[0];
+			data->acc[1] = (int16_t)buff[3];
+			data->acc[1] = (data->acc[1] * 256) + (int16_t)buff[2];
+			data->acc[2] = (int16_t)buff[5];
+			data->acc[2] = (data->acc[2] * 256) + (int16_t)buff[4];
+		} else {
+			LOG_WRN("Unhandled data: tag=%d", data_tag);
+		}
```

Comment:
Looking at this 'for' loop I'm lost. I also tried few years back to add FIFO support to lsm6dso, but I failed because seems that zephyr sensor APIs are currently lacking of this support, isn't it?

For example, here, it seems to me that you are reading all samples in FIFO, but then only the last one will be reported back thru API. Am I wrong?

Maybe I'm missing something...

Reply:
Yes that's right, there's not really any proper support on the sensor API that I could find. Hence, at the moment it only returns the last value, which means that the FIFO needs to be checked every time it's filled with one of each data_type if you want to read all values.

Looking again at the possibilities of the sensor API, I can see one way of adding proper support for the FIFO:
 - `sensor_sample_fetch_chan(dev, SENSOR_CHAN_ALL)` reads all data into a FIFO buffer
 - The call to `sensor_sample_fetch_chan` returns the size of the buffer - this would misuse the return value, which is usually 0 or a negative error code. Not sure if that's acceptable but it wouldn't break any other behaviour as far as I can see
 - The app can then loop over the buffer calling `sensor_channel_get(dev, SENSOR_CHAN_ALL, &sensor_value)` to retrieve the value together with the sensor channel type  of the current entry

Reply:
> Yes that's right, there's not really any proper support on the sensor API that I could find. Hence, at the moment it only returns the last value, which means that the FIFO needs to be checked every time it's filled with one of each data_type if you want to read all values.
> 
> Looking again at the possibilities of the sensor API, I can see one way of adding proper support for the FIFO:
> 
> * `sensor_sample_fetch_chan(dev, SENSOR_CHAN_ALL)` reads all data into a FIFO buffer
> * The call to `sensor_sample_fetch_chan` returns the size of the buffer - this would misuse the return value, which is usually 0 or a negative error code. Not sure if that's acceptable but it wouldn't break any other behaviour as far as I can see
> * The app can then loop over the buffer calling `sensor_channel_get(dev, SENSOR_CHAN_ALL, &sensor_value)` to retrieve the value together with the sensor channel type  of the current entry

Maybe you can add a request for API enhancement, which will be discussed by the steering committee. I think it is quite important.

Reply:
> Looking again at the possibilities of the sensor API, I can see one way of adding proper support for the FIFO:
> 
> * `sensor_sample_fetch_chan(dev, SENSOR_CHAN_ALL)` reads all data into a FIFO buffer
> * The call to `sensor_sample_fetch_chan` returns the size of the buffer - this would misuse the return value, which is usually 0 or a negative error code. Not sure if that's acceptable but it wouldn't break any other behaviour as far as I can see
> * The app can then loop over the buffer calling `sensor_channel_get(dev, SENSOR_CHAN_ALL, &sensor_value)` to retrieve the value together with the sensor channel type  of the current entry

Or we can add a brand new API, say sensor_channel_fifo_get() that passes a a data array and length parameter. THis was what I had in mind couple of years ago. But your solution looks maybe simpler.